File: api-testing-portfolio/playwright-api/data/booking.factory.ts
export interface BookingDates {
  checkin: string;
  checkout: string;
}

export interface BookingPayload {
  firstname: string;
  lastname: string;
  totalprice: number;
  depositpaid: boolean;
  bookingdates: BookingDates;
  additionalneeds?: string;
}

let sequence = 0;

function isoDate(daysFromNow: number): string {
  const date = new Date();
  date.setDate(date.getDate() + daysFromNow);
  return date.toISOString().slice(0, 10);
}

/**
 * Builds a unique booking per call. Restful Booker is a shared public
 * instance, so unique last names let tests filter for their own data and
 * never depend on a hardcoded booking ID.
 */
export function buildBooking(overrides: Partial<BookingPayload> = {}): BookingPayload {
  sequence += 1;
  return {
    firstname: 'Elene',
    lastname: `PortfolioTest-${Date.now()}-${sequence}`,
    totalprice: 100 + Math.floor(Math.random() * 900),
    depositpaid: true,
    bookingdates: {
      checkin: isoDate(30),
      checkout: isoDate(35),
    },
    additionalneeds: 'Breakfast',
    ...overrides,
  };
}

File: api-tests/src/test/java/booker/data/BookingFactory.java
package booker.data;

import booker.models.Booking;
import booker.models.BookingDates;

import java.time.LocalDate;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicInteger;

public class BookingFactory {
    private static final AtomicInteger seq = new AtomicInteger(1);

    public static Booking build() {
        return build(null);
    }

    // pass a booking with only the fields you want to override
    public static Booking build(Booking overrides) {
        LocalDate checkin = LocalDate.now().plusDays(30);
        LocalDate checkout = checkin.plusDays(5);

        Booking b = new Booking();
        b.firstname = "Elene";
        b.lastname = "PortfolioTest-" + System.currentTimeMillis() + "-" + seq.getAndIncrement();
        b.totalprice = ThreadLocalRandom.current().nextInt(100, 1000);
        b.depositpaid = true;
        b.bookingdates = new BookingDates(checkin.toString(), checkout.toString());
        b.additionalneeds = "Breakfast";

        if (overrides != null) {
            if (overrides.firstname != null) b.firstname = overrides.firstname;
            if (overrides.lastname != null) b.lastname = overrides.lastname;
            if (overrides.totalprice != null) b.totalprice = overrides.totalprice;
            if (overrides.depositpaid != null) b.depositpaid = overrides.depositpaid;
            if (overrides.bookingdates != null) b.bookingdates = overrides.bookingdates;
            if (overrides.additionalneeds != null) b.additionalneeds = overrides.additionalneeds;
        }
        return b;
    }
}
